### PR TITLE
Fix consumption of CPU-backed interchange protocol dataframes

### LINKED ISSUE #11245: [FEA] Raise informative error when interchanging non-cudf dataframes
If one were to [interchange](https://data-apis.org/dataframe-protocol/latest/API.html) with non-cudf dataframes via `cudf.core.df_protocol.from_dataframe()`, we'd get an error which doesn't quite relate to the problem at hand—that cross-device interchange isn't supported/isn't fully specified.

```python
>>> import numpy as np
>>> import vaex
>>> df = vaex.from_items(("foo", np.asarray([7, 42, 3])))
>>> df
  #    foo
  0      7
  1     42
  2      3
>>> from cudf.core.df_protocol import from_dataframe
>>> from_dataframe(df)
AttributeError: '_VaexBuffer' object has no attribute '_allow_copy'
```

<details>
<summary>Full traceback</summary>

```python
AttributeError                            Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 cudf_df = from_dataframe(df)

.../cudf/core/df_protocol.py:658, in from_dataframe(df, allow_copy)
    655 if not hasattr(df, "__dataframe__"):
    656     raise ValueError("`df` does not support __dataframe__")
--> 658 return _from_dataframe(df.__dataframe__(allow_copy=allow_copy))

.../cudf/core/df_protocol.py:681, in _from_dataframe(df)
    673 col = df.get_column_by_name(name)
    675 if col.dtype[0] in (
    676     _DtypeKind.INT,
    677     _DtypeKind.UINT,
    678     _DtypeKind.FLOAT,
    679     _DtypeKind.BOOL,
    680 ):
--> 681     columns[name], _buf = _protocol_to_cudf_column_numeric(col)
    683 elif col.dtype[0] == _DtypeKind.CATEGORICAL:
    684     columns[name], _buf = _protocol_to_cudf_column_categorical(col)

.../cudf/core/df_protocol.py:717, in _protocol_to_cudf_column_numeric(col)
    715 assert buffers["data"] is not None, "data buffer should not be None"
    716 _dbuffer, _ddtype = buffers["data"]
--> 717 _check_buffer_is_on_gpu(_dbuffer)
    718 cudfcol_num = build_column(
    719     Buffer(_dbuffer.ptr, _dbuffer.bufsize),
    720     protocol_dtype_to_cupy_dtype(_ddtype),
    721 )
    722 return _set_missing_values(col, cudfcol_num), buffers

.../cudf/core/df_protocol.py:728, in _check_buffer_is_on_gpu(buffer)
    725 def _check_buffer_is_on_gpu(buffer: _CuDFBuffer) -> None:
    726     if (
    727         buffer.__dlpack_device__()[0] != _Device.CUDA
--> 728         and not buffer._allow_copy
    729     ):
    730         raise TypeError(
    731             "This operation must copy data from CPU to GPU. "
    732             "Set `allow_copy=True` to allow it."
    733         )
    735     elif buffer.__dlpack_device__()[0] != _Device.CUDA and buffer._allow_copy:

AttributeError: '_VaexBuffer' object has no attribute '_allow_copy'
```
</details>

We get this `AttributeError` because [`_check_buffer_is_on_gpu()`](https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/python/cudf/cudf/core/df_protocol.py#L725) wants to raise an error when non-CuDF dataframes are being interchanged, but ends up assuming behaviour from CuDF dataframes (specifically buffers) first by using the non-specified attribute `_allow_copy`

https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/python/cudf/cudf/core/df_protocol.py#L725-L733

The method does raise an informative error which I think fits the bill here

https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/python/cudf/cudf/core/df_protocol.py#L735-L739

so I imagine desired behaviour here would be to just raise this if `buffer` doesn't say instance check to `_CuDFBuffer`.

Assumedly 1) data apis consortium will settle on behaviour 2) cudf adopts it, and thus this is a temporary issue :sweat_smile: A fix could be nice in the meantime still, but I namely submit this issue to keep track of things.

### PR DESCRIPTION
**Pull request by shwina**
## Description

Closes #11245. This PR fixes a bug in our code that consumes a protocol DataFrame that is backed by CPU memory.

This enables using the `from_dataframe()` function to construct DataFrames from other libraries:

```python
In [12]: pdf = pd.DataFrame({'x': [1, 2, 3], 'y': [1.0, 2.0, np.nan], 'z': ['a', 'b', 'c']})

In [13]: vdf = vaex.from_dict({'x': [1, 2, 3], 'y': [1.0, 2.0, np.nan]})

In [14]: cudf.from_dataframe(pdf, allow_copy=True)
Out[14]: 
   x    y  z
0  1  1.0  a
1  2  2.0  b
2  3  NaN  c

In [15]: cudf.from_dataframe(vdf, allow_copy=True)
Out[15]: 
   x    y
0  1  1.0
1  2  2.0
2  3  NaN
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by shwina** [comment]
rerun tests

File: python/cudf/cudf/tests/test_df_protocol.py [hunk 1]
```diff
@@ -232,3 +238,13 @@ def test_NA_mixed_dtype():
...
+
+
+@pytest.mark.skipif(
+    not PANDAS_GE_150,
+    reason="Pandas versions < 1.5.0 do not support interchange protocol",
+)
+def test_from_cpu_df(pandas_df):
+    df = pd.DataFrame({"a": [1, 2, 3], "b": ["x", "y", "z"]})
+    with pytest.raises(NotImplementedError):
+        cudf.from_dataframe(df, allow_copy=True)
```
Comment (wence-) [inline python/cudf/cudf/tests/test_df_protocol.py:250]: Can we assert this actually worked?

File: python/cudf/cudf/core/df_protocol.py [hunk 2]
```diff
@@ -660,13 +660,8 @@ def from_dataframe(
     if not hasattr(df, "__dataframe__"):
         raise ValueError("`df` does not support __dataframe__")
 
-    return _from_dataframe(df.__dataframe__(allow_copy=allow_copy))
+    df = df.__dataframe__(allow_copy=allow_copy)
```
Comment (wence-) [inline python/cudf/cudf/core/df_protocol.py:663]: Although `from_dataframe` previously didn't document the `allow_copy` argument, maybe now is a good time to do so?

**Comment by wence-** [comment]
Tests failed in `strings_udf` 
```
tables.pyx:13: in strings_udf._lib.tables.get_character_flags_table_ptr
    ???
E   RuntimeError: Fatal CUDA error encountered at: /workspace/.conda-bld/work/cpp/src/strings/utilities.cu:96: 222 unrecognized error code unrecognized error code
```

I lost track of what was happening there, will that be fixed by a back-merge from trunk?

**Review by wence- — CHANGES_REQUESTED** [review]
Very minor docstring quibble, otherwise looks good. I suppose the copies happen implicitly when the various `Buffer` objects are created?

File: python/cudf/cudf/tests/test_df_protocol.py (hunk 1 above)
Comment (shwina) [reply python/cudf/cudf/tests/test_df_protocol.py]: The goal of this PR was to raise the correct error message when consuming CPU-backed DataFrames. After reading your comment, I figured we might as well implement the bits needed to actually consume them. In the process I discovered a slew of issues that I've tried to address in this PR. I've re-requested your review, but I'll leave comments inline where I think it would be useful.

File: python/cudf/cudf/core/df_protocol.py [hunk 3]
```diff
@@ -46,6 +48,14 @@ class _Device(enum.IntEnum):
     ROCM = 10
 
 
+class _MaskKind(enum.IntEnum):
+    NON_NULLABLE = (0,)
+    NAN = (1,)
+    SENTINEL = (2,)
+    BITMASK = (3,)
+    BYTEMASK = 4
```
Comment (shwina) [inline python/cudf/cudf/core/df_protocol.py:56]: Prior to this PR, we used raw ints to represent this information.

File: python/cudf/cudf/core/df_protocol.py [hunk 4]
```diff
@@ -399,32 +408,22 @@ def _get_validity_buffer(
 
         Raises RuntimeError if null representation is not a bit or byte mask.
         """
-
```
Comment (shwina) [inline python/cudf/cudf/core/df_protocol.py:402]: It turns out that we were erroneously returning a boolean mask while promising that we were returning a bitmask (the interchange protocol allows returning either; the `col.describe_null` method tells the consumer which one to expect).

I've fixed this method to return a bitmask instead.

File: python/cudf/cudf/core/df_protocol.py [hunk 5]
```diff
@@ -645,7 +644,13 @@ def __dataframe__(
 _INTS = {8: cp.int8, 16: cp.int16, 32: cp.int32, 64: cp.int64}
 _UINTS = {8: cp.uint8, 16: cp.uint16, 32: cp.uint32, 64: cp.uint64}
 _FLOATS = {32: cp.float32, 64: cp.float64}
-_CP_DTYPES = {0: _INTS, 1: _UINTS, 2: _FLOATS, 20: {8: bool}}
+_CP_DTYPES = {
+    0: _INTS,
+    1: _UINTS,
+    2: _FLOATS,
+    20: {8: bool},
+    21: {8: cp.uint8},
```
Comment (shwina) [inline python/cudf/cudf/core/df_protocol.py:652]: Strings columns in the protocol are composed of a `uint8` buffer of `data` (chars) and an `int` buffer of `offsets`. The `data` column has a distinct type ID (`21`) that we previously didn't handle.

As a clarifying example, look at the buffers returned by a Pandas dataframe containing a single string column:

```python
In [17]: pd.DataFrame({'a': ['x', 'y', 'z']}).__dataframe__().get_column(0).get_buffers()
Out[17]: 
{'data': (PandasBuffer({'bufsize': 3, 'ptr': 140001666253824, 'device': 'CPU'}),
  (<DtypeKind.STRING: 21>, 8, 'u', '=')),
 'validity': (PandasBuffer({'bufsize': 3, 'ptr': 94467671186912, 'device': 'CPU'}),
  (<DtypeKind.BOOL: 20>, 8, 'b', '=')),
 'offsets': (PandasBuffer({'bufsize': 32, 'ptr': 94467669930400, 'device': 'CPU'}),
  (<DtypeKind.INT: 0>, 64, 'l', '='))}
```

File: python/cudf/cudf/tests/test_df_protocol.py [hunk 6]
```diff
@@ -7,25 +7,54 @@
...
 from cudf.testing._utils import assert_eq
 
 
+@pytest.fixture
+def pandas_df():
+    return pd.DataFrame({"a": [1, 2, 3], "b": ["x", "y", None]})
+
+
+def assert_validity_equal(protocol_buffer, cudf_buffer, size, null, valid):
+    if null == _MaskKind.BYTEMASK:
```
Comment (shwina) [inline python/cudf/cudf/tests/test_df_protocol.py:31]: Comparing boolmasks is easy, just use `assert_eq`.

File: python/cudf/cudf/tests/test_df_protocol.py [hunk 7]
```diff
@@ -7,25 +7,54 @@
...
+        protocol_mask = _protocol_buffer_to_cudf_buffer(protocol_buffer)
+        assert_eq(
+            as_column(protocol_mask, dtype="bool"),
+            as_column(cudf_buffer, dtype="bool"),
+        )
+    elif null == _MaskKind.BITMASK:
+        protocol_mask = _protocol_buffer_to_cudf_buffer(protocol_buffer)
```
Comment (shwina) [inline python/cudf/cudf/tests/test_df_protocol.py:38]: Comparing bitmasks is harder because we don't have utilities in Python to compare two bitmasks AFAICT. Instead, we construct two string columns composed entirely of empty strings, assign them the "left" and "right" bitmasks respectively, and then use `assert_eq` to compare the resulting columns.

Open to other suggestions here :-)

Comment (bdice) [reply python/cudf/cudf/tests/test_df_protocol.py:39]: If we construct a column of empty strings, it still requires offsets to be allocated/generated (which are `size_type = int32`). A more lightweight approach would be to use an `int8` data type with this mask.

File: python/cudf/cudf/core/df_protocol.py (hunk 2 above)
Comment (bdice) [reply python/cudf/cudf/core/df_protocol.py:668]: +1, please add parameter/return docstrings to `from_dataframe`.

**Review by bdice — CHANGES_REQUESTED** [review]
@shwina This popped back up on my radar after you bumped the branch. If this is still active, I think it only needs a bit more work to be complete?

File: python/cudf/cudf/core/df_protocol.py (hunk 3 above)
Comment (wence-) [reply python/cudf/cudf/core/df_protocol.py:56]: Is it deliberate that `BYTEMASK` is an int, but the other mask kinds are tuples? (They both turn into ints on construction, but the tuple instantiation looks kind of weird.

File: python/cudf/cudf/core/df_protocol.py [hunk 8]
```diff
@@ -719,49 +725,56 @@ def _protocol_to_cudf_column_numeric(
...
+                "This operation must copy data from CPU to GPU. "
+                "Set `allow_copy=True` to allow it."
+            )
+        else:
+            dbuf = rmm.DeviceBuffer(ptr=buf.ptr, size=buf.bufsize)
+            return _CuDFBuffer(
+                as_buffer(dbuf, exposed=True),
+                protocol_dtype_to_cupy_dtype(data_type),
+                allow_copy,
+            )
```
Comment (wence-) [inline python/cudf/cudf/core/df_protocol.py:752]: Style, I (at least) find double-negated conditions hard to parse. Perhaps have the truthy case in the if, and the exceptional case in the else branch:
```
if allow_copy:
    dbuf = ...
    return ...
else:
    raise TypeError(...)
```

**Review by wence- — COMMENTED** [review]
Some minor nitpicks, but otherwise looks good.

**Comment by vyasr** [comment]
@shwina what's the status of this PR?

File: python/cudf/cudf/core/df_protocol.py (hunk 2 above)
Comment (shwina) [reply python/cudf/cudf/core/df_protocol.py:700]: Done

File: python/cudf/cudf/core/df_protocol.py (hunk 3 above)
Comment (shwina) [reply python/cudf/cudf/core/df_protocol.py]: Fixed to just use ints here.

File: python/cudf/cudf/tests/test_df_protocol.py (hunk 7 above)
Comment (shwina) [reply python/cudf/cudf/tests/test_df_protocol.py:39]: Fixed to use an int8 column here instead

File: python/cudf/cudf/core/df_protocol.py (hunk 8 above)
Comment (shwina) [reply python/cudf/cudf/core/df_protocol.py]: Done

**Comment by shwina** [comment]
Thanks @vyasr and apologies all for letting this PR go stale - I've updated the PR to target 23.06, and I believe addressed all the review comments.

@bdice, @wence-, would you be willing to take a final look? Thank you for the suggested improvements!

**Review by wence- — COMMENTED** [review]
Implementation looks good, a minor request to add a little more testing in the pandas<->cudf interop tests

File: python/cudf/cudf/core/df_protocol.py [hunk 9]
```diff
@@ -828,27 +871,32 @@ def _protocol_to_cudf_column_string(
...
-    return _set_missing_values(col, cudfcol_str), buffers
+    return _set_missing_values(col, cudfcol_str, allow_copy), buffers
+
+
+def _protocol_buffer_to_cudf_buffer(protocol_buffer):
+    return as_buffer(
+        rmm.DeviceBuffer(
+            ptr=protocol_buffer.ptr, size=protocol_buffer.bufsize
+        ),
+        exposed=True,
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/df_protocol.py:901]: Not exactly. The relationship between spilling and copy-on-write is that enabling spilling and copy-on-write simultaneously is disallowed.

[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf/cudf/core/df_protocol.py:901]: I think what I meant was, in the same way that the buffer is exposed (disallowing spilling) if we take an external reference, it should also be marked as "exposed" such that a copy-on-write of the buffer copies eagerly.

[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/df_protocol.py:901]: Ah I see. Yes, I believe marking the buffer `exposed` will also set the `copy_on_write` flag appropriately. @galipremsagar can you confirm?

File: python/cudf/cudf/tests/test_df_protocol.py [hunk 10]
```diff
@@ -229,3 +268,12 @@ def test_NA_mixed_dtype():
...
     assert_df_unique_dtype_cols(data_mixed)
+
+
+@pytest.mark.skipif(
+    not PANDAS_GE_150,
+    reason="Pandas versions < 1.5.0 do not support interchange protocol",
+)
+def test_from_cpu_df(pandas_df):
+    df = pd.DataFrame({"a": [1, 2, 3], "b": ["x", "y", "z"]})
+    cudf.from_dataframe(df, allow_copy=True)
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/tests/test_df_protocol.py]: Thanks! I added some null tests.

> Also can we check the round-trip in the opposite direction? (Assuming that is supported too).

Today, that segfaults because Pandas doesn't check for the device type of the incoming buffers and tries to interpret the buffer pointers as host pointers :(

File: python/cudf/cudf/core/df_protocol.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/df_protocol.py:899]: Yup, this will mark it exposed and copy-on-write buffers will be copied eagerly. This change looks good 👍

File: python/cudf/cudf/core/df_protocol.py [hunk 11]
```diff
@@ -645,28 +644,61 @@ def __dataframe__(
...
+    dataframe interchange protocol (``__dataframe__``).
+
+    Parameters
+    ----------
+    df: Object supporting dataframe interchange protocol
+    allow_copy
+        If ``True``, allow copying of the data. If ``False``, a
+        ``TypeError`` is raised if data copying is required to
+        construct the ``DataFrame`` (e.g., if ``df`` lives in CPU
+        memory).
```
Comment (bdice) [inline python/cudf/cudf/core/df_protocol.py:670]: Fix the docstring syntax.
```suggestion
    df : DataFrameObject
        Object supporting dataframe interchange protocol
    allow_copy : bool
        If ``True``, allow copying of the data. If ``False``, a
        ``TypeError`` is raised if data copying is required to
        construct the ``DataFrame`` (e.g., if ``df`` lives in CPU
        memory).
```

**Review by bdice — APPROVED** [review]
One minor change for docstring syntax, otherwise LGTM.

**Review by wence- — APPROVED** [review]
Thanks, apologies for the delay
